fission-web-server/library/Fission/Web/Server/App/Modifier.hs

```diff
@@ -8,18 +9,135 @@ import           Database.Persist                                   as Persist
+import qualified Fission.Web.Server.Domain.Retriever.Class          as Domain.Retriever
 import           Fission.Web.Server.Error.ActionNotAuthorized.Types
+import qualified Fission.Web.Server.IPFS.DNSLink                    as DNSLink
 import           Fission.Web.Server.Models
-import           Fission.Web.Server.MonadDB
+import           Fission.Web.Server.MonadDB.Class (MonadDB(..))
+import           Fission.Web.Server.MonadDB.Types (Transaction)
 import           Fission.Web.Server.Ownership
 
+
+{-| Adds a file to an app.
+
+Notes:
+* Overwrites existing file data
+* Uses IPFS MFS as a temporary storage to manipulate the DAG structure
+
+-}
+addFile ::
+  ( App.Domain.Retriever m
+  , Domain.Retriever.Retriever m
+  , DNSLink.MonadDNSLink m
+  , MonadDB t m
+  , MonadLogger m
+  , MonadReader Config m
+  , MonadRemoteIPFS m
+  , MonadTime m
+  , Modifier t
+  , Retriever m
+  )
+  => UserId
+  -> DirectoryName
+  -> FileName
+  -> Lazy.ByteString
+  -> m (Either Errors' ())
+addFile userId appName@(DirectoryName appNameText) fileName rawData = do
+  now           <- currentTime
+  domainName    <- asks baseAppDomain
+
+  let subdomain = Subdomain appNameText
+  let url = URL domainName (Just subdomain)
+
+  mayApp        <- App.Retriever.byURL userId url
+
+  case mayApp of
+    Left _ -> do
+      return . Error.openLeft $ NotFound @App
+
+    Right (Entity appId app@App {appCid}) ->
+      if isOwnedBy userId app then do
+        -- Update DAG, DNSLink & DB
+        let (DomainName domainNameTxt) = domainName
+
+        appendToDag domainNameTxt appName fileName rawData appCid >>= \case
+          Left err ->
+            return $ Left err
+
+          Right newCID ->
+            IPFS.Stat.getSizeRemote newCID >>= \case
+              Left err ->
+                return . Error.openLeft $ err
+
+              Right size -> do
+                Domain.getByDomainName domainName >>= \case
+                  Left err ->
+                    return . Error.openLeft $ err
+
+                  Right Domain {domainZoneId} -> do
+                    DNSLink.set userId url domainZoneId newCID >>= \case
+                      Left err -> return . Error.openLeft $ err
+                      Right _  -> do
+                        runDB (setCIDDirectly now appId size newCID)
+                        return $ Right ()
+
+      else
+        return . Error.openLeft $ ActionNotAuthorized @App userId
+
+
+appendToDag ::
+  ( MonadRemoteIPFS m
+  , MonadLogger m
+  )
+  => Text
+  -> DirectoryName
+  -> FileName
+  -> Lazy.ByteString
+  -> CID
+  -> m (Either Errors' CID)
+appendToDag domainName (DirectoryName appName) (FileName fileName) rawData appCid = do
+  let appCidText = unaddress appCid
+
+  let tmpDirPath = Text.concat [ "/", domainName, "/", appName, "/" ]
+  let distDirPath = Text.concat [ tmpDirPath, appCidText, "/" ]
+  let filePath = Text.concat [ distDirPath, fileName ]
```

Comment:
We may want to write the files to an `uploads` subdirectory in the app. The original rationale is to have a dedicated location for files that are appended, but we should check if that's still what we want.

Reply:
Fixed 👍 I haven't allowed nested directories, but that should be as simple as allowing that to be passed to Servant (ie. now it doesn't allow multiple segments, could change that). I think we should probably allow that?

Reply:
@icidasset

No need to change, but FYI you can also do this with `<>`

```haskell
let filePath = distDirPath <> fileName
```

Reply:
Changed 👍